src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,268 @@
+            e.g. `(torch.optim.Adam, {'lr' : 1e-3})`.
+        qf_lr (float): Learning rate for Q-value network parameters.
+        clip_rewards (float): Clip reward to be in [-clip_rewards,
+            clip_rewards]. If None, rewards are not clipped.
+        clip_gradient (float): Clip gradient norm to `clip_gradient`. If None,
+            gradient are not clipped. Defaults to 10.
+        reward_scale (float): Reward scale.
+    """
+
+    def __init__(
+            self,
+            env_spec,
+            policy,
+            qf,
+            replay_buffer,
+            exploration_policy=None,
+            eval_env=None,
+            qf_optimizer=torch.optim.Adam,
+            *,  # Everything after this is numbers.
+            steps_per_epoch=20,
+            n_train_steps=50,
+            max_episode_length_eval=None,
+            buffer_batch_size=64,
+            min_buffer_size=int(1e4),
+            discount=0.99,
+            qf_lr=_Default(1e-3),
+            clip_rewards=None,
+            clip_gradient=10,
+            target_update_freq=5,
+            reward_scale=1.):
+        self._clip_reward = clip_rewards
+        self._clip_grad = clip_gradient
+
+        self._steps_per_epoch = steps_per_epoch
+        self._target_update_freq = target_update_freq
+        self._episode_qf_losses = []
+        self._epoch_ys = []
+        self._epoch_qs = []
+
+        self._policy = policy
+        self._qf = qf
+        self._n_train_steps = n_train_steps
+
+        self._min_buffer_size = min_buffer_size
+        self._qf = qf
+        self._steps_per_epoch = steps_per_epoch
+        self._n_train_steps = n_train_steps
+        self._buffer_batch_size = buffer_batch_size
+        self._discount = discount
+        self._reward_scale = reward_scale
+        self.max_episode_length = env_spec.max_episode_length
+        self._max_episode_length_eval = (max_episode_length_eval
+                                         or self.max_episode_length)
+        self._episode_reward_mean = collections.deque(maxlen=100)
+
+        self.env_spec = env_spec
+        self.replay_buffer = replay_buffer
+        self.policy = policy
+        self.exploration_policy = exploration_policy
+
+        self._target_qf = copy.deepcopy(self._qf)
+        self._qf_optimizer = make_optimizer(qf_optimizer,
+                                            module=self._qf,
+                                            lr=qf_lr)
+        self._eval_env = eval_env
+        self.worker_cls = FragmentWorker
+
+    def train(self, trainer):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            trainer (Trainer): Experiment trainer.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        if not self._eval_env:
+            self._eval_env = trainer.get_env_copy()
+        last_returns = [float('nan')]
+        trainer.enable_logging = False
```

Comment:
what is the purpose of this flag? Just curious as I haven't seen it in other algorithms.

Reply:
An example would be if you want to train and evaluate using different reward scales. For example, the `EpisodicLife` wrapper makes it so that episodes end after the agent loses a life, not until all lives are lost. This is fine for training, but if we wanted to report our episodic reward to be the total reward accumulated when **all** lives were lost, we would have to pass a clone of the environment to `eval_env` with the `EpisodicLife` wrapper removed. This would be helpful when comparing an agent's score to that of a human player, which is what deepmind does.

Reply:
Can you explain more?  I'm not sure I understand why you'd want to disable logging in this use case.

What you describe (passing the env you actually want evaluated to eval_env) is the garage way of achieving what you're describing. The intention is that the user will pass that modified env in the launcher (rather than the algo magically doing it).

Reply:
Ah, I was referring to `eval_env`. I disable logging initially to avoid logging before the warm up phase ends, ie. when the replay buffer contains `min_buffer_size` steps. All other off policy algos in garage do this. Not doing so will produce logs for the initial epochs where no training occurred, and also causes dowel to spit out a long warning about dynamically changing the logging keys. Ex:
```
2020-10-19 15:16:25 | [dqn_atari] Obtaining samples...
2020-10-19 15:16:28 | [dqn_atari] epoch #0 | Training finished
2020-10-19 15:16:50 | [dqn_atari] epoch #0 | Saving snapshot...
2020-10-19 15:16:50 | [dqn_atari] epoch #0 | Saved
2020-10-19 15:16:50 | [dqn_atari] epoch #0 | Time 24.99 s
2020-10-19 15:16:50 | [dqn_atari] epoch #0 | EpochTime 24.99 s
---------------------------  ------------
EpsilonGreedyPolicy/Epsilon      0.980042
TotalEnvSteps                10080
---------------------------  ------------
2020-10-19 15:16:50 | [dqn_atari] epoch #1 | Evaluating policy
2020-10-19 15:17:13 | [dqn_atari] epoch #1 | Training finished
2020-10-19 15:17:45 | [dqn_atari] epoch #1 | Saving snapshot...
2020-10-19 15:17:45 | [dqn_atari] epoch #1 | Saved
2020-10-19 15:17:45 | [dqn_atari] epoch #1 | Time 79.59 s
2020-10-19 15:17:45 | [dqn_atari] epoch #1 | EpochTime 54.60 s
/home/mishari/miniconda3/envs/garage/lib/python3.7/site-packages/dowel/logger.py:207: CsvOutputWarning: Inconsistent TabularInput keys detected. CsvOutput keys: {'EpsilonGreedyPolicy/Epsilon', 'TotalEnvSteps'}. TabularInput keys: {'Evaluation/StdReturn', 'QFunction/AverageY', 'Evaluation/AverageDiscountedReturn', 'Evaluation/MaxReturn', 'QFunction/AverageAbsQ', 'Evaluation/Iteration', 'Evaluation/100EpRewardMean', 'Evaluation/TerminationRate', 'Evaluation/AverageReturn', 'QFunction/MaxQ', 'QFunction/AverageQFunctionLoss', 'Evaluation/MinReturn', 'EpsilonGreedyPolicy/Epsilon', 'Epoch', 'TotalEnvSteps', 'Evaluation/NumEpisodes', 'QFunction/AverageAbsY', 'QFunction/AverageQ', 'QFunction/MaxY'}. Did you change key sets after your first logger.log(TabularInput)?
  output.record(data, prefix=self._prefix_str)
----------------------------------  -------------
Epoch                                   1
EpsilonGreedyPolicy/Epsilon             0.960083
Evaluation/100EpRewardMean            -20.3
Evaluation/AverageDiscountedReturn     -1.56143
Evaluation/AverageReturn              -20.3
Evaluation/Iteration                   20
Evaluation/MaxReturn                  -18
Evaluation/MinReturn                  -21
Evaluation/NumEpisodes                 10
Evaluation/StdReturn                    1.00499
Evaluation/TerminationRate              1
QFunction/AverageAbsQ                   0.869504
QFunction/AverageAbsY                   0.873958
QFunction/AverageQ                      0.866967
QFunction/AverageQFunctionLoss          0.0127423
QFunction/AverageY                      0.871055
QFunction/MaxQ                          1.53793
QFunction/MaxY                          1.86628
TotalEnvSteps                       20160
----------------------------------  -------------
```

Reply:
you can get around this by passing the number of timesteps that you need to fill the replay buffer to obtain episodes.